File: api/internal/domain/photo.go
package domain

import (
	"time"
)

// PhotoStatus represents the status of a photo
type PhotoStatus string

const (
	PhotoStatusDraft     PhotoStatus = "draft"
	PhotoStatusPublished PhotoStatus = "published"
)

type Photo struct {
	ID           uint        `gorm:"primaryKey" json:"id"`
	Title        string      `gorm:"size:200;not null" json:"title"`
	Description  string      `gorm:"type:text" json:"description"`
	ImageURL     string      `gorm:"size:500;not null" json:"imageUrl"`
	ThumbnailURL string      `gorm:"size:500" json:"thumbnailUrl"`
	Category     string      `gorm:"size:50" json:"category"`
	Location     string      `gorm:"size:200" json:"location"`
	IsFeatured   bool        `gorm:"default:false" json:"isFeatured"`
	DisplayOrder int         `gorm:"default:0;index" json:"displayOrder"`
	Status       PhotoStatus `gorm:"size:20;default:'draft';index;check:status IN ('draft','published')" json:"status"`
	CreatedAt    time.Time   `json:"createdAt"`
	UpdatedAt    time.Time   `json:"updatedAt"`
}

type CreatePhotoRequest struct {
	Title        string `json:"title" binding:"required,min=1,max=200"`
	Description  string `json:"description"`
	ImageURL     string `json:"imageUrl" binding:"required"`
	ThumbnailURL string `json:"thumbnailUrl"`
	Category     string `json:"category"`
	Location     string `json:"location"`
	IsFeatured   bool   `json:"isFeatured"`
	DisplayOrder int    `json:"displayOrder"`
}

type UpdatePhotoRequest struct {
	Title        *string      `json:"title"`
	Description  *string      `json:"description"`
	ImageURL     *string      `json:"imageUrl"`
	ThumbnailURL *string      `json:"thumbnailUrl"`
	Category     *string      `json:"category"`
	Location     *string      `json:"location"`
	IsFeatured   *bool        `json:"isFeatured"`
	DisplayOrder *int         `json:"displayOrder"`
	Status       *PhotoStatus `json:"status"`
}

// HasUpdates checks if the update request has at least one field to update
func (r *UpdatePhotoRequest) HasUpdates() bool {
	return r.Title != nil || r.Description != nil || r.ImageURL != nil ||
		r.ThumbnailURL != nil || r.Category != nil || r.Location != nil ||
		r.IsFeatured != nil || r.DisplayOrder != nil || r.Status != nil
}

File: web/lib/types/photo.ts
/**
 * 照片状态枚举
 * 对应后端 PhotoStatus
 */
export type PhotoStatus = "draft" | "published";

/**
 * 照片实体
 * 对应后端 domain.Photo
 */
export interface Photo {
  id: number;
  title: string;
  description: string;
  imageUrl: string;
  thumbnailUrl: string;
  category: string;
  location: string;
  isFeatured: boolean;
  displayOrder: number;
  status: PhotoStatus;
  createdAt: string; // ISO 8601 时间字符串
  updatedAt: string;
}

/**
 * 创建照片请求
 * 对应后端 CreatePhotoRequest
 */
export interface CreatePhotoRequest {
  title: string;
  description?: string;
  imageUrl: string;
  thumbnailUrl?: string;
  category?: string;
  location?: string;
  isFeatured?: boolean;
  displayOrder?: number;
}

/**
 * 更新照片请求
 * 对应后端 UpdatePhotoRequest
 * 所有字段都是可选的,只更新提供的字段
 */
export interface UpdatePhotoRequest {
  title?: string;
  description?: string;
  imageUrl?: string;
  thumbnailUrl?: string;
  category?: string;
  location?: string;
  isFeatured?: boolean;
  displayOrder?: number;
  status?: PhotoStatus;
}

/**
 * 组件照片关联
 * 对应后端 ComponentPhoto
 */
export interface ComponentPhoto {
  id: number;
  componentName: string;
  photoId: number;
  order: number;
  props: ComponentPhotoProps;
  photo: Photo;
  createdAt: string;
  updatedAt: string;
}

/**
 * 组件照片的自定义属性
 */
export interface ComponentPhotoProps {
  caption?: string;
  alt?: string;
  link?: string;
  [key: string]: any; // 允许扩展其他属性
}

/**
 * 创建组件照片关联请求
 */
export interface CreateComponentPhotoRequest {
  componentName: string;
  photoId: number;
  order: number;
  props?: ComponentPhotoProps;
}

/**
 * API 响应包装器
 */
export interface ApiResponse<T> {
  data: T;
  message?: string;
}

/**
 * 分页响应
 */
export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  pageSize: number;
}
